crates/iota-aws-orchestrator/src/orchestrator.rs

```diff
@@ -320,14 +322,230 @@ impl<P: ProtocolCommands<T> + ProtocolMetrics, T: BenchmarkType> Orchestrator<P,
             .wait_for_command(self.instances(), id, CommandStatus::Terminated)
             .await?;
 
-        // Execute and wait for the cargo build command on all instances except the
-        // metrics one. This requires compiling the codebase in release
-        // (which may take a long time) so we run the command in the background
-        // to avoid keeping alive many ssh connections for too long.
+        // Define the binaries to build/download
+        let binaries = vec![
+            "iota".to_string(),
+            "iota-node".to_string(),
+            "stress".to_string(),
+        ];
+
+        // Check if build cache is enabled
+        if let Some(build_config) = &self.settings.build_cache_server {
+            if build_config.enabled {
+                display::action("Using build cache for binary distribution");
+                self.update_with_build_cache(build_config, commit, &binaries)
+                    .await?;
+            } else {
+                self.update_with_local_build(&binaries).await?;
+            }
+        } else {
+            self.update_with_local_build(&binaries).await?;
+        }
+
+        display::done();
+        Ok(())
+    }
+
+    /// Detect CPU target architecture for all instances that need binaries.
+    /// Returns a HashMap mapping CPU target to list of instances with that
+    /// target
+    async fn detect_cpu_targets_for_instances(
+        &self,
+    ) -> TestbedResult<HashMap<String, Vec<Instance>>> {
+        let instances_needing_binaries = self
+            .instances()
+            .iter()
+            .filter(|i| i.role != InstanceRole::Metrics)
+            .cloned()
+            .collect::<Vec<_>>();
+
+        if instances_needing_binaries.is_empty() {
+            return Ok(HashMap::new());
+        }
+
+        display::action("Detecting CPU targets for all instances");
+
+        let context = CommandContext::new();
+        let command = "rustc --print target-cpus".to_string();
```

Comment:
```suggestion
        let command = "source \"$HOME/.cargo/env\" && rustc --print target-cpus".to_string();
```

I'm testing this feature, and this command was stuck because rustc wasn't in the `PATH` for non-interactive SSH sessions. So we need to source the envs.